src/application_settings/container_base.py

```diff
@@ -0,0 +1,267 @@
+        return new_settings
+
+    @classmethod
+    def _create_instance(cls: type[_ContainerT]) -> _ContainerT:
+        """Instantiate the Container."""
+
+        # get whatever is stored in the config/settings file
+        data_stored = cls._get_stored_data()
+        # filter out fields that are both stored and an attribute of the Container
+        _data_fields = {
+            fld for fld in fields(cls) if fld.init and fld.name in data_stored.keys()
+        }
+        # instantiate the sections and keep them in a dict
+        # actually sections: dict[str, _ContainerSectionT]
+        # but MyPy doesn't swallow that
+        sections: dict[str, Any] = {
+            fld.name: cls._instantiate_section(fld.type, data_stored[fld.name])
+            for fld in _data_fields
+        }
+
+        # instantiate the Container with the sections
+        return cls(**sections)
+
+    @classmethod
+    def _get_stored_data(cls) -> dict[str, Any]:
+        """Get the data stored in the toml file"""
+        data_stored: dict[str, Any] = {}
+
+        if path := cls.filepath():
+            if (ext := path.suffix[1:]) == FileFormat.TOML.value:
+                with path.open(mode="rb") as fptr:
+                    data_stored = tomllib.load(fptr)
+            elif ext == FileFormat.JSON.value:
+                with path.open(mode="r") as fptr:
+                    data_stored = json.load(fptr)
+            else:
+                print(f"Unknown file format {path.suffix[1:]} given in {path}.")
+        else:
+            # This situation can occur if no valid path was given as an argument, and
+            # the default path is set to None.
+            print(
+                f"No path specified for {cls.kind_string().lower()}file; trying with defaults, but this may not work."
+            )
+        return data_stored
+
+    @classmethod
+    def _instantiate_section(
+        cls: type[_ContainerT],
+        class_to_instantiate: type[_ContainerSectionT],
+        arg_dict: dict[str, Any],
+    ) -> _ContainerSectionT:
+        """Return an instance of class_to_instantiate, properly initialized"""
+        # pre-condition: class_to_instantiate is the class of an initializable field of cls
+        assert (
+            len([f for f in fields(cls) if f.init and f.type == class_to_instantiate])
+            > 0
+        )
+
+        field_set = {f.name for f in fields(class_to_instantiate) if f.init}
+        filtered_arg_dict = {k: v for k, v in arg_dict.items() if k in field_set}
+        return class_to_instantiate(**filtered_arg_dict)
+
+    def _set(self) -> None:
+        """Private method to store the singleton."""
+        _ALL_CONTAINERS[id(self.__class__)] = self
+
+    def _save(self) -> None:
+        """Private method to save the singleton to file."""
+        if path := self.filepath():
+            path.parent.mkdir(parents=True, exist_ok=True)
+            if (ext := path.suffix[1:]) == FileFormat.TOML.value:
```

Comment:
I think eventually it might be nice to abstract the reading and writing from this class. So idealy we would just have `def read(path: Path) -> dict[str, Any]`: and `def write(path: Path, data: dict[str, Any]) -> bool:`, these methods could then do the actual wrapping of expected library.

Reply:
Created an issue for this.